### ISSUE TITLE
Score is different for same parameter in different version (0.9.1 vs master)

### ISSUE THREAD

**Issue by papperwing** [deeplearning4j/deeplearning4j#4286]
After update I have started to get weird score for my NN so I tested if same behavior is on totaly same parameters on 0.9.1 and here are graphs:

master:
![snimek obrazovky 20](https://user-images.githubusercontent.com/5627436/32697779-89f5a14a-c798-11e7-8d09-b8fe815a9265.png)

0.9.1:
![snimek obrazovky 22](https://user-images.githubusercontent.com/5627436/32697781-91266eb8-c798-11e7-97e6-9cf41acaa44b.png)

Recently there have been change in computing l1/l2 in frozen layer could that affect whole computing of gradient or propagation of updates?

**Comment by AlexDBlack** [comment]
Certainly, the l1/l2 fix for compgraph + frozen layers did affect the gradients/updates.

@papperwing I assume I can pull the latest version here to reproduce this?
https://github.com/papperwing/image-net

**Comment by papperwing** [comment]
@AlexDBlack Yes You just have to change dl4j.version property in top-most pom.xml and in ModelBuilderImpl you must change way the learning rate is set when changing version to 0.9.1. Nothing else is necessary to change.

**Comment by papperwing** [comment]
I tried to run learning without freezing layers and it still behaves differently. So it will not be caused by frozen layer, because they shouldn't be there.
![snimek obrazovky 24](https://user-images.githubusercontent.com/5627436/32731933-6a057e7a-c88b-11e7-8980-e4d5343e5903.png)

**Comment by AlexDBlack** [comment]
I can confirm that there is definitely *something* going on here... left is 0.9.1, right is master (run on Linux)

![image](https://user-images.githubusercontent.com/2360237/32813274-6198e722-c9fd-11e7-87ed-ed4b35124331.png)

**Comment by AlexDBlack** [comment]
Related, and possibly the cause: https://github.com/deeplearning4j/deeplearning4j/issues/4302

**Comment by AlexDBlack** [comment]
Update: https://github.com/deeplearning4j/deeplearning4j/issues/4302 is fixed, but unfortunately this does not appear to be the cause :(
I'm still seeing a difference between 0.9.1 and master here.

**Comment by raver119** [comment]
Is difference reproducible for both cpu & cuda?

**Comment by raver119** [comment]
and cuda with cudnn?

**Comment by AlexDBlack** [comment]
I've only been testing with CPU currently. I spent a bunch of time last night looking into this, haven't isolated it yet. I'm focused on this "nearly flat score" issue I'm seeing currently.

It's not a  workspace/scope issue as far as I can tell. Params and updater state looks fine. Scores and activations look about the same for both.
But if I load the 0.9.1 net into master, I get quite different results during training.

**Comment by raver119** [comment]
Try CUDA with cuDNN please.

**Comment by AlexDBlack** [comment]
Success - at least for the flat score issue...
Turns out that was dropout(0) actually appling dropout with 0 retain probability.
https://github.com/deeplearning4j/deeplearning4j/pull/4326
https://github.com/deeplearning4j/deeplearning4j/pull/4326/commits/7e03ccfe54b95a8a9a3987c40e1b7f74b0c1f6f9
https://github.com/deeplearning4j/deeplearning4j/pull/4326/commits/bcc7f1ad7bd3fd7c5951f14cf0d1ccb81787f1a2

I'll run this longer to see if it behaves sensibly, before closing.

**Comment by raver119** [comment]
O_o

**Comment by AlexDBlack** [comment]
Yeah, definitely seems to be learning properly now...
![image](https://user-images.githubusercontent.com/2360237/33259864-528a935c-d3b2-11e7-92f9-4edd6f93129b.png)

**Comment by AlexDBlack** [comment]
https://github.com/deeplearning4j/deeplearning4j/pull/4326
